### PR TITLE
Allow users to edit workbench GPU to 0

### LINKED ISSUE #881: [Bug]: Editing Workbench Does Not Respect Zero GPUs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When we are in edit mode of an existing workbench within a DS Project -- setting zero GPUs is actually considered not doing anything with GPUs... whereas if the workbench ever had GPUs enabled, it would then never be able to remove the GPU setting from the value it was last set to zero.

### Expected Behavior

We should look to clear the GPU setting when GPUs are set to zero if they are already on a workbench.

### Steps To Reproduce

1. Have a Workbench with a configured GPU setting
2. Edit the Workbench in the DS Project UI
3. Set the GPU value to 0
4. Save the Workbench with no other changes

-- we should have zero configured at this point on this workbench --

### Workaround (if any)

_No response_

### What browsers are you seeing the problem on?

_No response_

### Open Data Hub Version

```yml
2.5.2
```

### Anything else

For development clusters without GPUs, you could possibly use the `gpuSetting` value in the DashboardConfig to fake GPUs -- your pod likely won't start, but it should be an effective test (which we will need to verify on cluster at some point).

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
## Description

Closes #881 

## How Has This Been Tested?

Since I don't have a GPU cluster, I only did some fake tests and checked the Notebook CR, it works.
1. Go to the dashboard config CR, set `notebookController -> gpuSetting: '3'`
2. Go to the spawner page, you will see the GPU number dropdown
3. Select the GPU number
4. Go to check the Notebook CR
5. Edit the workbench and set the GPU to 0
6. Check the Notebook CR again

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by maroroman** [comment]
running retest to verify if tests are failing across repo

/retest

File: frontend/src/api/network/notebooks.ts [hunk 1]
```diff
@@ -256,8 +258,14 @@ export const updateNotebook = (
...
   const notebook = assembleNotebook(data, username);
 
+  const container = existingNotebook.spec.template.spec.containers[0];
+
   // clean the envFrom array in case of merging the old value again
-  existingNotebook.spec.template.spec.containers[0].envFrom = [];
+  container.envFrom = [];
+  // clean the resources, affinity and tolerations for GPU
+  existingNotebook.spec.template.spec.tolerations = [];
+  existingNotebook.spec.template.spec.affinity = {};
```
Comment (andrewballantyne) [inline frontend/src/api/network/notebooks.ts:267]: We need to clone this. I missed it during the rapid development of DSG. You're mutating likely state from the context here -- we can't trust variable sources of what is passed to us, let us avoid mutating those.

File: frontend/src/api/network/notebooks.ts [hunk 2]
```diff
@@ -43,6 +43,8 @@ const assembleNotebookAffinityAndTolerations = (
       operator: 'Exists',
     });
   } else {
+    delete resources.limits?.['nvidia.com/gpu'];
+    delete resources.requests?.['nvidia.com/gpu'];
```
Comment (andrewballantyne) [inline frontend/src/api/network/notebooks.ts:47]: Ah yeah -- probably shouldn't mutate this either -- can we return an updated set of resources or something instead?

Mutations of variables you are passed has a long history of causing issues weirdly. Imagine a bug in a scenario completely separate of this, 2 months from now. When I do xyz the data just doesn't appear anymore.... or is different than what I did... and I have no idea why. Turns out, it was some function call in a dozen that mutated the variable. Best to always be pure about input so you don't have a headache later.

File: frontend/src/api/network/notebooks.ts [hunk 3]
```diff
@@ -256,12 +262,19 @@ export const updateNotebook = (
   data.notebookId = existingNotebook.metadata.name;
   const notebook = assembleNotebook(data, username);
 
+  const oldNotebook = { ...existingNotebook };
```
Comment (andrewballantyne) [inline frontend/src/api/network/notebooks.ts:265]: You'll need to use `_.cloneDeep` unfortunately... spreads are shallow copies and thus when you dig into properties like `.spec` that's an object reference -- so you are now mutating the previous object.

File: frontend/src/types.ts [hunk 4]
```diff
@@ -68,10 +68,12 @@ export type NotebookResources = {
   requests?: {
     cpu?: string;
     memory?: string;
+    'nvidia.com/gpu'?: number;
```
Comment (andrewballantyne) [inline frontend/src/types.ts:71]: These are strings in my case 🤔 I wonder if we should make sure to covert to number from the override config setting. It worked on clusters with GPUs I think -- so we should be _fine_ for now.

Could you define a "special type" for GPU?
```
/** 
 * OdhDashboardConfig contains gpuSetting as a string value override -- proper gpus return as numbers
 * TODO: Look to make it just number by properly parsing the value
 */
type GPUCount = string | number;
```
```suggestion
    'nvidia.com/gpu'?: GPUCount;
```

That way we can document it once and look to change it in the future.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Code works good -- minor request to change the type.

**Review by andrewballantyne — APPROVED** [review]

**Pull request merged**